modules/custom/falcon_commerce/src/Plugin/rest/resource/DonationResource.php

```diff
@@ -0,0 +1,294 @@
+    parent::__construct($configuration, $plugin_id, $plugin_definition, $serializer_formats, $logger);
+    $this->entityTypeManager = $entity_type_manager;
+    $this->time = $time;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->getParameter('serializer.formats'),
+      $container->get('logger.factory')->get('falcon_commerce'),
+      $container->get('entity_type.manager'),
+      $container->get('datetime.time')
+    );
+  }
+
+  /**
+   * Handle POST requests.
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   Request object.
+   *
+   * @return \Drupal\rest\ResourceResponse
+   *   Response object.
+   */
+  public function post(Request $request) {
+    $this->logger->info($request->getContent());
+
+    // TODO: check if we can get this data some other way provided by REST.
+    // Look into the serialization_class.
+    $this->data = json_decode($request->getContent(), TRUE);
+
+    $order = $this->createOrder();
+
+    // TODO: add payment processing.
+    // $this->>processPayment($order);
+
+    $response = new ResourceResponse($order, 200);
+
+    return $response;
+  }
+
+  /**
+   * Creates an order from given data.
+   */
+  protected function createOrder() {
+    $store_ids = [];
+    $order_items = [];
+    $variation_sku = 'donation';
+    /** @var \Drupal\commerce_product\Entity\ProductVariationInterface $variation */
+    $variations = $this->entityTypeManager->getStorage('commerce_product_variation')->loadByProperties(['sku' => $variation_sku]);
+    $variation = reset($variations);
+
+    if (!$store_ids) {
+      $product = $variation->getProduct();
+      $store_ids = $product->getStoreIds();
+    }
+
+    // Create order item object.
+    /** @var \Drupal\commerce_order\Entity\OrderItemInterface $order_item */
+    $order_item = $this->entityTypeManager->getStorage('commerce_order_item')->create([
+      'title' => $this->t('Donation'),
+      'type' => 'donation',
+      'purchased_entity' => $variation,
+      'quantity' => 1,
+      'field_donation_type' => $this->data['donation_type'],
+      'appeal' => $this->data['appeal'],
+    ]);
+
+    // Set price.
+    $unit_price = new Price($this->data['payment']['amount'], $this->data['payment']['currency_code']);
+    $order_item->setUnitPrice($unit_price, TRUE);
+
+    try {
+      // We need only entity level violations.
+      /** @var \Drupal\Core\Entity\EntityConstraintViolationListInterface $violations */
+      $violations = $order_item->validate()->getEntityViolations();
+      if ($violations->count() > 0) {
+        foreach ($violations as $violation) {
+          $this->logger->info($violation->getPropertyPath() ?: 'Entity' . ': ' . $violation->getMessage());
```

Comment:
Could you please increase the severity of this log message?

Reply:
Set it to warning.